### DISCUSSION TITLE
Version 4.0.0

### DISCUSSION THREAD

**Discussion by piotrooo** [asterisk-java/asterisk-java#487] [Show and tell]
@scgm11 maybe you're not familiar with this, but I've done small [road map](https://github.com/asterisk-java/asterisk-java/issues/405), could you look at this?

Furthermore, I've been started to refactor couple of this (repackaged, extracted AMI, refactored sockets) [here](https://github.com/asterisk-java/asterisk-java/tree/feature/4.x.x).
One of the big thing is a cleanup databind'er which could be connected with an issue #338.

I would appreciate it if you look at this and give some feedback.

I also have plan to move from Maven to Gradle.

**Comment by scgm11** [comment]
First of all thanks for the work on this project as it was almost  abandoned.

1. Agree with repackaged and decouple modules
2. yes that is an important issue to address, I remember also a long time ago to loose some events so I stopped updating was something to do with https://github.com/asterisk-java/asterisk-java/issues/262 I think the FastScanner gave me issues but I dont remember exactly.
3. yes Gradle would be much better!
4. Github Actions and Releases is the way to go, also continue publishing on maven central 
5. Yes all should be done with feature branches and merge them with PRs as we did today
6. Your proposed module structure looks good to me.
7. Ok with JAVA (Im currently starting to use 19 previously 8) 
8. Im ok with the version structure (Im currently starting to use asterisk 20 previously 13)
9. Logo looks amazing!
10. Agree with everything you said!

**Reply by piotrooo** [reply]
Now when you join to the project I have a little bit more motivation to do next steps 😉

**Reply by piotrooo** [reply]
Maybe you want do join with help? We can describe tasks to do and do it together.

**Reply by scgm11** [reply]
yes, Im just retaking a project that will use this, also Im just going back to code a bit .. so I can take some stuff after I finish to do some work on the project, I can also help testing the new version structure and if you change the reflection part.

**Comment by scgm11** [comment]
Im not sure about the labeling kind/.... maybe go directly with bug, new feature, improvemnt, and others and use the colors @piotrooo what do you think?

**Reply by piotrooo** [reply]
I've been took this idea from e.g. https://github.com/go-gitea/gitea/labels?q=kind or https://github.com/spring-projects/spring-framework/labels?q=type

Maybe `type: *` is better, but I strongly recommend to use prefix to describe type of the label.

**Reply by scgm11** [reply]
ok no problem for me

**Comment by scgm11** [comment]
hey @piotrooo can we make a release merging CDR and MES I think are good candidates for the release in v3
Code is tested and working

**Reply by piotrooo** [reply]
Sure! There should be a 3.39.0 version 😉

**Comment by seanbright** [comment]
Maybe we can discuss here instead of the other discussion?

**Comment by piotrooo** [comment]
@seanbright I finally finished a project, and I should have a bit of free time to start working on this one. Apologies for the long delay in my response 🙏.

I have a few ideas about the future and the direction of the project development.

1. Migrate from Maven to Gradle
For me, Gradle feels more natural. Many of the largest open-source communities use Gradle (such as Spring and OpenTelemetry).

2. Modularization Asterisk Java
Introduce several modules, such as `agi`, `ami`, `ari`, and `core`, to create clear, well-defined boundaries.

3. Introduce integration testing using Testcontainers
We need to create a robust Asterisk Docker image to facilitate integration testing with Testcontainers. Without that type of testing, we can't be sure if our library really works. It's crucial! This Docker image should be stored in a separated repository within the asterisk-java organization.

4. Introduce Netty
As long as we handle everything ourselves, we are not focusing on developing features for the Asterisk Java project. Instead, we are concentrating on low-level programming aspects like reading from sockets. A library like Netty can handle this for us, and do it in a better, more optimized way.

5. Introduce new serializers and deserializers
While working with JacksonXML, I created several serializers and deserializers that leverage JacksonXML features for marshalling and unmarshalling payloads sent to Asterisk. You can find them here: https://github.com/asterisk-java/asterisk-java/tree/577-move-all-actions-and-responses-to-ami/asterisk-java-core/src/main/java/org/asteriskjava/core/databind

Those are my ideas. Let me know what you think about them. Right now, with support, it's possible to accomplish these tasks.

**Reply by seanbright** [reply]
Before I respond to this, does it supersede [the proposed roadmap](https://github.com/asterisk-java/asterisk-java/issues/405) you created previously?

**Reply by piotrooo** [reply]
> Before I respond to this, does it supersede [the proposed roadmap](https://github.com/asterisk-java/asterisk-java/issues/405) you created previously?

I think it enriches the roadmap. This list should work with the roadmap.

**Reply by seanbright** [reply]
I don't have a ton of time right now but I'll address a couple of the items while I do:

> 1. Migrate from Maven to Gradle
> For me, Gradle feels more natural. Many of the largest open-source communities use Gradle (such as Spring and OpenTelemetry).

There would have to be some pretty compelling benefits to using Gradle for me to consider moving away from Maven. I could swap "Maven" for "Gradle" in your sentence above (and swap in a few large OSS projects using Maven) and that would be my justification for sticking with Maven.

> 2. Modularization ...

Sure.

> 3. Introduce integration testing using Testcontainers

I don't feel strongly about this (and it doesn't sound like something I would want to work on), but if it's something you wanted to spearhead, go for it.

> 4. Introduce Netty

That's fine. It would be useful on the AGI, ARI (future), and AudioSocket (future) sides. Not so much on the AMI side but no reason we couldn't use it there too.

> 5. Introduce new serializers and deserializers

I need some more time to review this code to understand what you are proposing. This replaces the AMI event/action/response parsing I assume?

A few minor things I want to do for v4 are:

1. Remove our home-grown logging abstraction in favor of slf4j's
2. Potentially auto-generate AMI actions/events from Asterisk XML documentation.
3. ~~There was a 3rd but it disappeared from my brain before I could write it down. Will update later.~~ Change master to main. I think I remember seeing someone else (you?) mention that elsewhere.

**Reply by piotrooo** [reply]
> There would have to be some pretty compelling benefits to using Gradle for me to consider moving away from Maven.

I mentioned that the Gradle feels more natural to me. Personally, I don't like XML. Gradle is much cleaner than the XML in Maven, which creates a lot of noise. I don't have any deal breakers, though I prefer Gradle.

> Remove our home-grown logging abstraction in favor of slf4j's

It sounds extremely appealing. Totally agree.

> Potentially auto-generate AMI actions/events from Asterisk XML documentation.

I'm not familiar with this concept. Could you share some additional info or links?

> Change master to main. I think I remember seeing someone else (you?) mention that elsewhere.

Yes, definitely!

What do you think we should start with? Maybe renaming `master` to `main` would be a goot first step. But what should we do next?

**Reply by seanbright** [reply]
>> Potentially auto-generate AMI actions/events from Asterisk XML documentation.
>
> I'm not familiar with this concept. Could you share some additional info or links?

There is [XML documentation embedded within the source code of Asterisk](https://github.com/asterisk/asterisk/blob/7f2776129d968116d7eb035220c6fcecf3414897/main/manager.c#L106-L1565) that is extracted and merged into an XML file (typically `/var/lib/asterisk/documentation/core-en_US.xml`) used by the asterisk process for the help system.

I'm suggesting leveraging that file to generate our action/event "models" (e.g. `org.asteriskjava.manager.action.GetVarAction`). There is very little, if any, type information in that file so my thought is that we would build something that turned `core-en_US.xml` into another format (yaml maybe) that we could then hand edit to add type information and then use _that_ file to generate the Java code.

> What do you think we should start with? Maybe renaming `master` to `main` would be a goot first step. But what should we do next?

Since we're in agreement I'll go ahead and change the branch name. Should we branch off a 3.x branch and start working on v4 in main? I think splitting into multiple modules would be a good first step after that.

**Reply by seanbright** [reply]
Actually we should probably wait to branch 3.x until we’re ready to release 4.0. I know you have a 4.x branch but would it make sense to start over since we’ll stick with Maven?
